Add coin input control

chimp1984 opened the pull request:
Adds input selection for BSQ send transactions. We use a utxo based model in contrast to the address based model in the Funds screen. If any dev wants to apply that to the funds screen as well would be good. Also it is better IMO to hide it in a popup as it is a bit technical and might confuse some users.

<img width="1198" alt="Screen Shot 2021-01-31 at 21 30 01" src="https://user-images.githubusercontent.com/54558767/106408773-1f744c80-640d-11eb-87a3-6929dd6080f9.png">

<img width="880" alt="Screen Shot 2021-01-31 at 21 29 41" src="https://user-images.githubusercontent.com/54558767/106408771-1c795c00-640d-11eb-9506-c0884c1e769b.png">

chimp1984 committed:
Extract onSendBsq method, return early

chimp1984 committed:
Add refreshValidation method

chimp1984 committed:
Implement isDustAttackUtxo protection

chimp1984 committed:
Add addButtonCheckBoxWithBox method with top param

chimp1984 committed:
Add support for utxo set used in coin selection if set
We use that to use only the selected utxos instead of all available.

chimp1984 committed:
Add getSpendableBsqTransactionOutputs and getSpendableNonBsqTransactionOutputs
methods.
Add overridden getPreparedSendBsqTx and getPreparedSendBtcTx methods
with utxoCandidates param.

If utxoCandidates is not null we apply it to our coinSelector.
As the coin selector is re-used we re-set it immediately after
it was applied (inside coin selector select method).

Set preferences in nonBsqCoinSelector

chimp1984 committed:
Add TxInputSelectionWindow for coin input selection.
It is both for BSQ and non-BSQ (BTC) used.

chimp1984 committed:
Merge branch 'master' of https://github.com/bisq-network/bisq into master_upstream

chimp1984 committed:
Merge branch 'master' of https://github.com/bisq-network/bisq into master_upstream

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/NonBsqCoinSelector.java:
```diff
@@ -60,9 +64,9 @@ protected boolean isTxOutputSpendable(TransactionOutput output) {
     }
 
-    // BTC utxo in the BSQ wallet are usually from rejected comp request so we don't expect dust attack utxos here.
+    // Prevent to use dust attack utxos
```
```suggestion
    // Prevent usage of dust attack utxos
```

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/BisqDefaultCoinSelector.java:
```diff
@@ -65,7 +74,16 @@ public BisqDefaultCoinSelector() {
+        ArrayList<TransactionOutput> sortedOutputs;
+        if (utxoCandidates != null) {
+            sortedOutputs = new ArrayList<>(utxoCandidates);
+            // We we reuse the selectors we reset the transactionOutputCandidates field
```
```suggestion
            // We reuse the selectors. Reset the transactionOutputCandidates field
```

sqrrm commented on desktop/src/main/java/bisq/desktop/main/dao/wallet/send/BsqSendView.java:
```diff
@@ -171,11 +188,16 @@ protected void activate() {
 
         bsqWalletService.addBsqBalanceListener(this);
 
+        // We reset the input selection at active to have all inputs selected, otherwise the user
```
```suggestion
        // We reset the input selection at activate to have all inputs selected, otherwise the user
```

sqrrm commented on desktop/src/main/java/bisq/desktop/main/dao/wallet/send/BsqSendView.java:
```diff
@@ -282,16 +322,49 @@ private void onSendBsq() {
+        if (bsqUtxoCandidates == null) {
+            bsqUtxoCandidates = new HashSet<>(unspentTransactionOutputs);
+        } else {
+            // If we had some selection stored we need to update to already spent entries
+            bsqUtxoCandidates = bsqUtxoCandidates.stream().
+                    filter(e -> unspentTransactionOutputs.contains(e)).
+                    collect(Collectors.toSet());
```
This would remove transactions that are no longer spendable BSQ outputs from the selected outputs. I think that could be confusing. Better to check if the `bsqUtxoCandidates` are still all spendable, if they are then all is good, if not, then reset the selection to allow choosing between all `unspentTransactionOutputs` again.

chimp1984 replied:
To request the spendable via bsqWalletService.getSpendableBsqTransactionOutputs() and then check if the bsqUtxoCandidate is in that set does check if the bsqUtxoCandidate is spendable. We use the coinSelector for that and there is no API for checking one utxo if it is still spendable.

Or do I misunderstand what you mean?

sqrrm commented on desktop/src/main/java/bisq/desktop/main/dao/wallet/send/BsqSendView.java:
```diff
@@ -306,43 +379,80 @@ private void addSendBtcGroup() {
+            btcUtxoCandidates = new HashSet<>(unspentTransactionOutputs);
+        } else {
+            // If we had some selection stored we need to update to already spent entries
+            btcUtxoCandidates = btcUtxoCandidates.stream().
+                    filter(e -> unspentTransactionOutputs.contains(e)).
+                    collect(Collectors.toSet());
```
Same comment as for the BSQ, could be confusing if the selection changes without the users knowing about it

chimp1984 replied:
If you make a send tx and spend a previous selected utxo we need to remove that. We always show that spendable utxos and those are marked the user has selected. A previous selected which got spent need to be removed.

sqrrm reviewed (commented):
Testing issue:
1. Start with a BSQ wallet with no BTC balance
2. BSQ selection popup shows when button is pressed
3. Send BTC to BSQ receive address (remove leading "B")
4. BTC balance section shows in BSQ wallet but neither "select inputs" button work.

chimp1984 committed:
Merge branch 'master' of https://github.com/bisq-network/bisq into master_upstream

chimp1984 committed:
Fix typos

chimp1984 committed:
Set onAction handlers in activate not in initialize

chimp1984 commented:
> BTC balance section shows in BSQ wallet but neither "select inputs" button work.

There was a bug with the button handlers. Fixed in b37dc15

chimp1984 committed:
Merge branch 'master' of https://github.com/bisq-network/bisq into master_upstream

chimp1984 committed:
Merge branch 'master' of https://github.com/bisq-network/bisq into master_upstream

chimp1984 committed:
Add comment

ripcurlx commented:
Would be great to reset the validation of the input fields similar as it was before, so it doesn't look like an error after coming back from the success popup.

<img width="1317" alt="Bildschirmfoto 2021-02-18 um 21 43 54" src="https://user-images.githubusercontent.com/170962/108419187-8037bd80-7232-11eb-89cd-5501003201d3.png">

chimp1984 committed:
Improve resetValidation method to reset validation result

If text field is empty we apply ValidationResult(true), otherwise we apply
the validate result with the text and the given validator.

sqrrm reviewed (approved):
ACK
